### PR TITLE
Make DummyRange work this types other than uint

### PR DESCRIPTION
**Pull request by burner**
* fixating the current DummyRange behaviour in a unittest.
* structured the tests a little
* DummyRange with custom types
* DummyRange line coverage from < 10% to 100%
* DummyRange and AllDummyRanges bahaviour have unchanged interface and semantics

### CODE REVIEW

File: std/internal/test/dummyrange.d [hunk 1]
```diff
@@ -28,23 +28,40 @@ enum ReturnBy
...
 }
 
+import std.traits : isArray;
+
 // Range that's useful for testing other higher order ranges,
 // can be parametrized with attributes.  It just dumbs down an array of
 // numbers 1..10.
-struct DummyRange(ReturnBy _r, Length _l, RangeType _rt)
+struct DummyRange(ReturnBy _r, Length _l, RangeType _rt, T = uint[])
+        if(isArray!T)
```
Comment (andralex) [inline std/internal/test/dummyrange.d]: space after if, sigh

File: std/internal/test/dummyrange.d [hunk 2]
```diff
@@ -28,23 +28,40 @@ enum ReturnBy
...
+        [1U, 2U, 3U, 4U, 5U, 6U, 7U, 8U, 9U, 10U];
     // These enums are so that the template params are visible outside
     // this instantiation.
     enum r = _r;
     enum l = _l;
     enum rt = _rt;
 
-    uint[] arr = [1U, 2U, 3U, 4U, 5U, 6U, 7U, 8U, 9U, 10U];
+    //static if (isArray!T && is(Unqual!(ElementType!T) == uint))
+    static if (is(T == uint[]))
```
Comment (andralex) [inline std/internal/test/dummyrange.d]: ... like here. Really, inconsistent formatting within the same PR cannot go through.

Comment (burner) [inline std/internal/test/dummyrange.d]: all if's and for(each)'s not are followed by an space

File: std/internal/test/dummyrange.d [hunk 3]
```diff
@@ -28,23 +28,41 @@ enum ReturnBy
...
+    }
+
+    alias RetType = ElementType!(T);
+    alias RetTypeNoAutoDecoding = ElementEncodingType!(T);
 
     void reinit()
     {
         // Workaround for DMD bug 4378
-        arr = [1U, 2U, 3U, 4U, 5U, 6U, 7U, 8U, 9U, 10U];
+        static if (is(T == uint[]))
```
Comment (9il) [inline std/internal/test/dummyrange.d]: How this would work for int?

Comment (9il) [inline std/internal/test/dummyrange.d]: Do `int` and `uint` dummy ranges have different behaviour?

Comment (burner) [inline std/internal/test/dummyrange.d]: The behavior is not different, but the original DummyRange was an uint[] only (1 to 11) range, and I wanted to preserve that bahavior so the DummyRange api would not break. But reinit broke every DummyRange(T) where a uint[] could not implicitliy be assigned to a T[]

File: std/internal/test/dummyrange.d [hunk 4]
```diff
@@ -271,4 +307,251 @@ unittest
...
+                copy[i] += Cmp.dummyValue;
+            }
+
+            assert(equal!(Cmp.cmp)(copy, Cmp.dummyValueRslt.repeat(copy.length)));
+        }
+    }
+
+    import std.meta : AliasSeq;
+
+    foreach (S; AliasSeq!(uint,double,TestFoo))
```
Comment (quickfur) [inline std/internal/test/dummyrange.d]: More style nitpicks: please put a space after `,`, here and also elsewhere in this PR. It makes the result easier to read.

**Comment by 9il** [comment]
> Wait, so this is adding tests for an internal unit test helper that's already being testing from other unit tests? Or am I missing something?

@burner added new functionality, it should be tested. In addition, it is an example for Phobos contributors.
D Style :)

**Comment by burner** [comment]
@klickverbot this module was used to test other modules, it's own behavior was not tested.
I added the possible to use different types as uint with it. Before I added to possibility to use different types but uint, I added tests to make sure I did not change some behavior unused in phobos.

**Comment by burner** [comment]
@9il I do not follow

**Comment by quickfur** [comment]
ping
Where are we going with this?  Since this is only updating internal test ranges, I'm inclined to just merge it.  Any objections?  @andralex @klickverbot @9il

**Comment by 9il** [comment]
LGTM

**Comment by burner** [comment]
as said earlier, the tests where added to fixate the api and behavior of the original DummyRange implementation, before I added the new functionality to DummyRange.

So it is not just tests for tests,

**Comment by 9il** [comment]
Auto-merge toggled on

**Comment by burner** [comment]
I would argue, that long term having DummyRange!T will yield better tests for other functions and it will fixing possible bugs in DummyRange easier.

I will write a test with iota to test Cmp(T) and TestFoo ...

**Comment by burner** [comment]
@9il thanks

**Pull request merged**